privacyidea/api/lib/postpolicy.py

```diff
@@ -318,6 +318,54 @@ def no_detail_on_success(request, response):
     return response
 
 
+def preferred_client_mode(request, response):
+    """
+    This policy function is used to add the preferred client mode.
+    The admin can set the list of client modes in the policy in the
+    same order as  he preferred them. The faction will pick the first
+    client mode from the list, that is also in the multichallege and
+    set it as preferred client mode
+
+    :param request:
+    :param response:
+    :return:
+    """
+    content = response.json
+
+    # get the preferred client mode from a policy definition
+    detail_pol = Match.action_only(g, scope=SCOPE.AUTH, action=ACTION.PREFERREDCLIENTMODE)\
+        .action_values(unique=True)
+
+    if detail_pol:
+        preferred_client_mode_list = detail_pol.get("test")[0].split(", ")
+    else:
+        preferred_client_mode_list = ['poll', 'webauthn', 'interactive', 'u2f']
+    if content.get("detail"):
+        detail = content.get("detail")
+        if detail.get("multi_challenge"):
+            multi_challenge = detail.get("multi_challenge")
+            l = []
+            for x in multi_challenge:
+                l.append(x.get("client_mode"))
+            l=list(dict.fromkeys(l))
```

Comment:
What is this for? I think if we wanted to reduce the list, we did it usually via a set:

    list(set(l))

On the other hand, since you are only testing if it is **in** `l`, then reducing the list is not necessary.

Reply:
[34ee918](https://github.com/privacyidea/privacyidea/pull/3361/commits/34ee918173549cdef4dfeb1e000951457fff1cab)